### PR TITLE
fix(ai,settings): defer CLI model probes and bound captured output

### PR DESCRIPTION
**Pull request by theBGuy**
Opening Settings → AI immediately probed the configured provider's model catalog, which for CLI providers means spawning the CLI as a subprocess before the user has touched the model picker. This gates the CLI arm behind actual engagement with the picker and hardens the capture helper underneath so a child process can't buffer output without limit.

### Deferred CLI probe (`src/features/settings/AiProviderSection.tsx`)

- `ModelPicker` now holds a sticky `modelsWanted` flag and passes `{ enabled: !isCli || modelsWanted }` to `useAvailableModels`, so a CLI provider's catalog is only listed once the user reaches the picker; HTTP providers keep fetching eagerly, since that path is a GET rather than a subprocess.
- The flag is armed from the `Combobox`'s `onOpenChange` (on open) and the `ComboboxInput`'s `onFocus`, covering both click and keyboard-focus entry.
- The hint switch reads `availableModels.isFetching` instead of `isPending` for the "Loading models…" and "Model passed to the CLI…" arms, matching the query's new placeholder-backed status.
- `isCli` moves above the `useAvailableModels` call, since it now feeds the enabled gate.

### Placeholder catalog (`src/lib/ai/models.ts`)

- `useAvailableModels` gains `placeholderData` returning `fallbackModels(settings)` with `live: false`, so the picker shows static suggestions while the probe runs and while the query is gate-disabled instead of an empty list — parity with the session pickers' `useAgentModels`.
- A comment records the consequence for consumers: `placeholderData` forces query status to `success`, so loading must be detected via `isFetching`.

### Bounded CLI capture (`src-tauri/src/agent.rs`)

- `run_capture_parts` replaces `cmd.output()` with an explicit `spawn()` plus a `tokio::select!` loop that drains stdout and stderr concurrently into separate buffers, so a full pipe can't deadlock the child.
- Adds `MAX_CAPTURE_BYTES` (8 MiB) as a per-stream cap; a stream that hits it triggers `child.start_kill()` and ends the loop, bounding memory as well as duration (the existing timeout bounded only the latter).
- The read loop and the following `child.wait()` remain inside the same `tokio::time::timeout`, still surfacing `AppError::Timeout`, with the exit code, stdout, and stderr now derived from the collected buffers.

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevented commands that produce excessive output from overwhelming the application by stopping captures beyond a safe limit.
  * Improved command cleanup and error handling when output exceeds the limit.
  * Output exactly at the safe limit is now handled correctly.

* **Improvements**
  * CLI model lists now load when the model selector is opened or focused, making settings load faster.
  * Added suggested models while live model information is loading or unavailable.
  * Improved loading-state messaging across model selectors during background catalog refreshes.
  * Clarified when catalogs are loading, when no matches are found, and that typed model IDs can still be used.

### CODE REVIEW

**Comment by theBGuy** [comment]
**Context for reviewers** — two residuals of #246 (opencode live model catalog), which merged after the v0.9.6 cut and so is unreleased.

**What & why**
1. The Settings → AI and PR-review model pickers listed nothing for up to ~20s during a cold `opencode models` probe, and opening Settings → AI with opencode selected spawned the CLI on mount. `useAvailableModels` now stands in its static suggestions via `placeholderData` (so the list is never empty), and a CLI provider's catalog probe waits for picker intent — parity with the session pickers (`useAgentModels`), which have done both since #246.
2. `run_capture_parts` buffered command output with no byte limit (the timeout bounded duration, not bytes). It now drains stdout/stderr concurrently under an 8 MiB per-stream cap and kills the child on overflow.

**Deliberate calls**
3. Loading detection is `isFetching`, not `isPending`. `placeholderData` forces the query status to `success` (so `isPending` is always false); `isFetching` is the correct in-flight signal, measured against query-core 5.101.4. PrReviewPanel already uses `isFetching` for exactly this reason.
4. Only the CLI catalog probe is intent-gated (`enabled: !isCli || modelsWanted`); HTTP providers keep fetching eagerly. Reaching Settings → AI is itself intent to configure AI, and an HTTP catalog fetch is a GET, not a subprocess — the concern was the CLI spawn on mount. This keeps every HTTP hint state, with no misleading pre-intent message.
5. The CLI `--version`/auth detection still runs on mount (the "Found (1.18.18)" status) — only the model *catalog* probe is deferred. Deferring detection would hide install status until interaction.
6. The 8 MiB cap is a safety valve far above any legitimate output (the largest, `opencode models`, is tens of KB). On overflow the child is killed and the truncated output returns a non-zero code, so callers fall back to suggestions — the spawned binaries are already trusted, so this is defense-in-depth, not a functional path. Normal-path capture is byte-identical to `Command::output`.
7. Live-verified in a cold-start build (opencode 1.18.18): selecting opencode showed the gated hint with no opencode process; focusing the picker spawned it and populated the live catalog; the popup was never empty; the hint settled to "7 models from opencode (CLI)".

**Disclosures**
8. No changelog fragment: both defects are post-v0.9.6-cut (#246 unreleased), and the drain is symptomless hardening — no user-observable released bug. `no-changelog` label applied.
9. `resolve_via_login_shell` (`src-tauri/src/agent.rs:641`) keeps its unbounded `.output()` — stdout-only, a trusted-shell `command -v` returning a path, timeout-bounded; negligible risk. `run_capture_parts` funnels the rest of the capture family.
10. No README/site/help change: the picker still fills from the live catalog and starts on the account default; the deferral is timing the docs never claimed.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI review** · `opus` · automated

---

## Summary

Two independent residuals of #246: the Settings model picker now stands in static suggestions via `placeholderData` and defers a CLI provider's catalog probe until the user reaches the picker, and `run_capture_parts` replaces `Command::output()` with a hand-rolled concurrent drain under an 8 MiB per-stream cap. Both are sound — the drain avoids the pipe-fill deadlock correctly (`select!` with per-branch `if !done` guards, `else => break` on dual EOF, `kill_on_drop` still covering the timeout path), and the `isPending` → `isFetching` switch is the right consequence of adding `placeholderData`. Nothing blocking; the gaps are test coverage on the new cap and the exported hook's contract doc.

### Tests

- **should-fix** — `src-tauri/src/agent.rs:715-761`, the new capped drain. The cap/kill branch is the one path no manual run will ever exercise, and it carries a boundary condition (`n >= room`) plus a truncating slice — yet it's untestable as written, because the loop is inlined in a function that spawns a real process and `MAX_CAPTURE_BYTES` is a fixed 8 MiB const. Concretely: nothing today would catch it if `n.min(room)` were dropped and `extend_from_slice(&obuf[..n])` overran the cap, or if the overflow `break` were placed before the `extend_from_slice`. Extract the loop into an inner helper that takes the cap as a parameter and reports overflow rather than killing:

  ```rust
  async fn capture_capped<O: AsyncRead + Unpin, E: AsyncRead + Unpin>(
      stdout: &mut O, stderr: &mut E, cap: usize,
  ) -> AppResult<(Vec<u8>, Vec<u8>, bool)>  // (out, err, overflowed)
  ```

  `run_capture_parts` then calls it with `MAX_CAPTURE_BYTES` and does the `child.start_kill()` itself when `overflowed`. `&[u8]` implements `AsyncRead`, so a `#[tokio::test]` (the file's `mod tests` at :3142 is sync-only today, but 31 modules in `src-tauri/src` already use `#[tokio::test]`) can drive it with `cap: 16` and assert: under-cap input round-trips byte-identically on *both* streams, over-cap input truncates to exactly `cap` with `overflowed == true`, and an exactly-`cap` input lands on whichever side of the boundary you settle below. Knock-on: `MAX_CAPTURE_BYTES`'s doc comment (:689-693) stays on the const, and the `// Read both pipes concurrently…` comment at :716-717 moves onto the helper, since that's where the deadlock constraint now lives.

### Correctness / edge cases

- **nit** — `src-tauri/src/agent.rs:735` and `:748`: `if n >= room` kills the child when the read *exactly* fills the remaining room, i.e. an output of precisely `MAX_CAPTURE_BYTES` is truncation-free but still reported as a failed run. Use `n > room` — it still terminates, because once `out.len() == MAX_CAPTURE_BYTES` any further byte gives `room == 0` and `n > 0`.
- **nit** — `src-tauri/src/agent.rs:736`/`:749` + `:757`: on overflow the killed status flows into `agent_models` (:862-869), which reports the first non-empty stderr line — so a capture that blew the cap surfaces to the user as an arbitrary CLI banner line, or `"opencode models exited with code -1"`, indistinguishable from a crashed binary. Returning `Err(AppError::Command(format!("{} produced more than {MAX_CAPTURE_BYTES} bytes of output", program.display())))` straight from `run_capture_parts` names the real cause; both callers already propagate `AppResult` (`run_capture` at :775 via `?`, `agent_detect` at :791 via `.ok()`), so no call site changes.

### Clarity / contract

- **should-fix** — `src/lib/ai/models.ts:231-236`, `useAvailableModels`'s doc comment. The new `placeholderData` changes the hook's consumer-visible contract — `data` is now never `undefined` (including while gate-disabled), status is always `success`, and `isPending` is permanently `false` — but that's recorded only in an inline comment on the option at :294-298, which a consumer reading the hook signature at a call site never sees. This is precisely the trap that produced half of this PR (`isPending` at `AiProviderSection.tsx:143`/`:145` silently never firing). Add it to the exported doc, e.g. after the existing "`opts.enabled` lets a caller defer…" sentence: *"`data` is always defined — static suggestions stand in until the request settles, and whenever the query is gate-disabled — so the query status is always `success`: detect an in-flight request with `isFetching`, never `isPending`, and `data.live`/`data.cause` to tell a settled fallback from a stand-in."* The sibling `useAgentModels` (:307-315) already ships `placeholderData` (:342) with the same undocumented contract and the same one-sentence fix; worth carrying in the same pass since the wording is shared.
- **nit** — `src/features/settings/AiProviderSection.tsx:145`: `case availableModels.isFetching` sits above `case catalog?.live === true`, so for an HTTP provider a background refetch (staleTime 5 min + `refetchOnWindowFocus: true`) replaces the settled "63 models from OpenAI" with "Loading models…". Moving the `live === true` case above the `isFetching` case fixes it without affecting first load or a provider switch — both change the query key, so `live` is `false` on the placeholder and the loading case still wins.
- **nit** — `src/features/settings/AiProviderSection.tsx:222-224`: `ComboboxEmpty` keeps the flat "No matching models — the typed id is used as-is", but with placeholder suggestions always present it now only renders under a typed filter, which may still be in-flight. The two siblings already signal this — `AgentPickers.tsx:114-116` (`available.isFetching ? "No matching models yet (catalog loading) — …" : …`) and `PrReviewPanel.tsx:416-418`; mirror the AgentPickers form here.

### Recorded decisions

Notes 3 (`isFetching` over `isPending`), 4 (CLI-only intent gate), 5 (detection stays eager), and 10 (no README/site/help change) are recorded and consistent with the code I read. Note 8's no-fragment call checks out independently: `changelog.d/added-opencode-live-models.md` is still pending, so #246 is unreleased and neither defect was observable in a shipped build.

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI security audit** · `opus` · automated

---

No exploitable security issues in these changes.

Notes on the other reviewers' items, both verifiable against the current diff and neither a security issue:

- Copilot (`agent.rs:735`, truncation returns `Ok` with a possibly-zero status): a robustness nit, not a vulnerability. Every caller of `run_capture`/`run_capture_parts` here (`--version`, `auth status`, `opencode models`, `docker/podman version|image inspect`) parses output from a binary the user already trusts to run, gates on `code == 0` for short outputs, and none of them approaches 8 MiB, so there is no attacker-controlled path from a truncated capture to a security decision.
- CodeRabbit (`n >= room` should be `n > room`): an off-by-one in the conservative direction — it kills at exactly the cap rather than over it. `out.len()` can never exceed `MAX_CAPTURE_BYTES`, so `MAX_CAPTURE_BYTES - out.len()` cannot underflow; no memory or correctness hazard beyond the noted edge case.

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._

File: src-tauri/src/agent.rs [hunk 1]
```diff
@@ -706,14 +712,57 @@ pub(crate) async fn run_capture_parts(
...
+                r = stdout.read(&mut obuf), if !odone => {
+                    let n = r.map_err(AppError::Io)?;
+                    if n == 0 {
+                        odone = true;
+                    } else {
+                        let room = MAX_CAPTURE_BYTES - out.len();
+                        out.extend_from_slice(&obuf[..n.min(room)]);
+                        if n >= room {
+                            let _ = child.start_kill();
+                            break;
```
[the remark this answers is not in the document]
Comment (theBGuy) [reply src-tauri/src/agent.rs]: Accepted — and it's the sharper version of context item 6: I claimed the kill yields a non-zero code, but in the race where the child already exited 0 (wrote past the cap, then finished), `child.wait()` returns the real `0`, so a caller would parse truncated output as complete success. Fixed in the next push: on cap-hit `run_capture_parts` returns `Err(AppError::Command("<prog> produced more than {MAX_CAPTURE_BYTES} bytes of output"))`, so every caller treats it as a deterministic failed probe (they fall back to suggestions). I also extracted the drain into a pure `capture_capped(stdout, stderr, cap) -> (out, err, overflowed)` helper — the caller does the `start_kill` — with `#[tokio::test]`s covering under-cap round-trip, over-cap truncation, and the exact-cap boundary.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

File: src/features/settings/AiProviderSection.tsx [hunk 2]
```diff
@@ -110,14 +110,20 @@ function ModelPicker({
   allowedHosts: string[];
 }) {
   const keyPreview = useSecretPreview(value.provider);
+  const isCli = isCliProvider(value.provider);
+  // Listing a CLI's catalog spawns it, so a CLI provider's probe waits for the
+  // user to reach the picker — sticky, matching the session and PR-review
+  // pickers. HTTP providers keep fetching eagerly: reaching Settings → AI is
+  // itself the intent, and the cost is a GET, not a subprocess.
+  const [modelsWanted, setModelsWanted] = useState(false);
```
[the remark this answers is not in the document]
Comment (theBGuy) [reply src/features/settings/AiProviderSection.tsx:118]: The `no-changelog` label — the exemption `changelog.d/README.md` names — is applied. Both defects are post-v0.9.6-cut (#246, the opencode live-models PR, is unreleased) and the capture-drain is symptomless hardening, so neither was observable in a shipped build. The changelog ships in-app as the release/updater notes for *released* user-facing bugs, so a fragment here would be a phantom entry.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

**Comment by theBGuy** [comment]
Thanks — all six items landed in the next push. Verified on the fix delta: `cargo test --lib` 1400/0 (3 new tests), `clippy --all-targets -D warnings` clean, `tsc -b`, biome ci on both TS files.

- **Cap helper + tests (should-fix):** extracted `capture_capped<O, E>(stdout, stderr, cap) -> AppResult<(Vec<u8>, Vec<u8>, bool)>` exactly as sketched; `run_capture_parts` does the `start_kill` when `overflowed`. Three `#[tokio::test]`s drive it with `cap: 16` — under-cap round-trips both streams byte-identically, over-cap truncates to 16 with `overflowed == true`, and exactly-cap keeps all 16 and is **not** flagged. The `MAX_CAPTURE_BYTES` doc stays on the const; the deadlock comment moved onto the helper.
- **`n >= room` → `n > room`:** done (in the helper), so output landing exactly on the cap is truncation-free — a later non-empty read (`room == 0`) flags real overflow.
- **Overflow → error:** `run_capture_parts` now returns `Err(AppError::Command("<prog> produced more than {MAX_CAPTURE_BYTES} bytes of output"))` on cap-hit — this supersedes context item 6's "non-zero code" (the exited-0 race made that false). `run_capture`/`agent_models` propagate via `?`, `agent_detect` via `.ok()`, so it's a clean fallback with no call-site changes.
- **Hook contract doc:** added your suggested sentence to `useAvailableModels`, and the shared one-liner to `useAgentModels`.
- **Hint order:** moved `catalog?.live === true` above `availableModels.isFetching`, with a comment on why (a background refetch keeps `live` data; first load / provider switch changes the query key so the loading case still wins).
- **`ComboboxEmpty`:** now mirrors the AgentPickers in-flight form.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI review** · `opus` · automated

---

Two residuals of #246 land here: the Settings model picker gates a CLI catalog probe behind picker intent and stands in static suggestions via `placeholderData`, and `run_capture_parts` now drains stdout/stderr concurrently under an 8 MiB per-stream cap, failing deterministically on overflow. The round's fixes are correct and I found no blocker — the spawn conversion in particular is safe because `stdin(null)`/`stdout(piped)`/`stderr(piped)` are set at `agent.rs:764-766`, so `expect("stdout piped")` can't fire even though `Command::output()` used to set piping implicitly, and every `run_capture`/`run_capture_parts` call site (`agent.rs:809/825/834/895/2261`, `health.rs:65/76`, `agent_sandbox.rs:242/249/256/510/525/773/929/978`) already treats `Err` benignly, so the new overflow error needs no call-site change.

### Tests

- **should-fix** — `src-tauri/src/agent.rs:3179-3206`, the three new `capture_capped` tests. All three drive stdout only (`(big.as_slice(), b"")`, `(exact.as_slice(), b"")`), so the stderr arm at `:732-744` — a hand-copied twin of the stdout arm, minus the explanatory comment — has zero coverage of its boundary. Concretely: flip its `n > room` to `n >= room`, or drop its `break`, and every test still passes while `capture_capped` returns `overflowed == false` for an over-cap stderr; `run_capture_parts` then returns `Ok` with a silently truncated stderr and an unkilled child — exactly the case the new `Err` path at `:787-795` exists to prevent. Add a fourth test that swaps the streams:

  ```rust
  #[tokio::test]
  async fn capture_capped_truncates_and_flags_over_cap_on_stderr() {
      let big = vec![b'x'; 100];
      let (mut o, mut e): (&[u8], &[u8]) = (b"", big.as_slice());
      let (out, err, overflowed) = capture_capped(&mut o, &mut e, 16).await.unwrap();
      assert!(out.is_empty());
      assert_eq!(err.len(), 16);
      assert!(overflowed);
  }
  ```

  Deterministic under `select!`'s random branch order: the empty stdout slice yields `n == 0` → `odone` whichever arm is polled first. No other change needed — the helper's signature, the `MAX_CAPTURE_BYTES` const doc, and the existing three tests all stay as-is.

### Clarity / contract

- **should-fix** — `src-tauri/src/agent.rs:751-753`, `run_capture_parts`'s doc comment. This push changed the function's error contract (overflow used to return `Ok` with truncated bytes; it now returns `Err(AppError::Command(...))` at `:787-795`), but the doc still describes only the happy path, so a caller reading the signature learns the cap exists only by scrolling to the const 60 lines up. Append one sentence to the existing doc:

  ```rust
  /// Runs a short command and returns (exit code, stdout, stderr) as separate
  /// streams — what a line parser needs, since a banner on stderr must not
  /// interleave into the data being parsed. Each stream is capped at
  /// `MAX_CAPTURE_BYTES`: a child that writes past it is killed and the call
  /// fails with `AppError::Command` rather than returning a truncated capture.
  ```

  Knock-on: none. `run_capture` (`:803`) delegates and propagates via `?`, so its one-line doc needs no edit, and `MAX_CAPTURE_BYTES`'s own doc stays on the const where this push put it.

### Recorded decisions

Notes 8 and 10 (no changelog fragment, no README/site/help change) remain recorded decisions and check out independently: `changelog.d/added-opencode-live-models.md` is still pending, so #246 is unreleased and neither defect was observable in a shipped build, and the one guide claim about this surface — `src/features/help/content.ts:2087-2089`, "its model picker fills from the CLI's own catalog … starting on your account's default model" — stays true under the deferral. Copilot's missing-fragment comment is answered by the `no-changelog` exemption `changelog.d/README.md` names; not a finding.

### Resolved since last review

- `capture_capped` extracted as a pure `AsyncRead`-driven helper with the cap as a parameter and overflow reported rather than killed (`agent.rs:695-749`), with `run_capture_parts` doing the `start_kill` (`:779-781`) — the deadlock comment moved onto the helper and the const doc stayed put, as asked.
- `n >= room` → `n > room` on both arms (`agent.rs:726`, `:739`), so output landing exactly on the cap is no longer reported as a failed run (also what CodeRabbit flagged).
- Overflow now returns `Err(AppError::Command(...))` (`agent.rs:787-795`), so the exited-0 race can no longer surface a truncated capture as a complete success (also what Copilot flagged).
- Hook contract documented on both `useAvailableModels` (`models.ts:235-239`) and `useAgentModels` (`:315-317`).
- Hint order: `catalog?.live === true` now precedes `availableModels.isFetching` (`AiProviderSection.tsx:150-153`), so a background refetch no longer flashes "Loading models…" over a settled count.
- `ComboboxEmpty` mirrors the AgentPickers in-flight form (`AiProviderSection.tsx:227-234`).

### Leftover polish (non-blocking)

- `src/features/pulls/PrReviewPanel.tsx:415-419`: adding `placeholderData` to the shared `useAvailableModels` means that panel's popup is never empty either (every `MODEL_SUGGESTIONS` entry is non-empty), so its `ComboboxEmpty` now renders only under a non-matching typed filter — where "Loading models…" describes the wrong cause. Mirror the two siblings: `available.isFetching ? "No matching models yet (catalog loading) — the typed id is used as-is" : "No matching models — the typed id is used as-is"`; the `isFetching`-never-`isPending` comment above it stays accurate.

Verdict: no blocking issues — remaining items are non-blocking; merge when ready

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._

**Comment by theBGuy** [comment]
Round-2 items all land in the next push. Verified: `cargo test --lib` 1401/0, `clippy --all-targets -D warnings` clean, `tsc -b`, biome ci on the edited file.

- **stderr-arm coverage (should-fix):** added your `capture_capped_truncates_and_flags_over_cap_on_stderr` as written — and one refinement to the coverage claim: an over-cap read alone can't tell `>` from `>=` (100 bytes against a 16 cap flags overflow under both operators), so the discriminating case for the stderr boundary is *exact-cap on stderr*. The exactly-at-cap test now drives both streams. Confirmed by mutation: flipping the stderr arm's `>` to `>=` turns the exact-cap test red (its stderr block) while both over-cap tests stay green.
- **`run_capture_parts` doc (should-fix):** appended your sentence verbatim — it checks out against the code (kill at `capture_capped` return, `AppError::Command` at the overflow gate).
- **Per-stream wording (CodeRabbit):** the overflow message now reads "produced more than {MAX_CAPTURE_BYTES} bytes on stdout or stderr" — the cap is per-stream, the old text implied a total.
- **PrReviewPanel `ComboboxEmpty` (polish):** mirrored to the sibling form. Accepted as in-batch rather than deferred — it's the third instance of the same class this PR (my round-1 sweep stopped at the Settings picker; a class sweep should have caught this one then).

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI review** · `opus` · automated

---

Two round-2 items and CodeRabbit's per-stream wording all land correctly, and the fix hunks themselves are clean: the new stderr test is order-independent under `select!` (the empty stdout slice yields `n == 0` → `odone` whichever arm polls first), and the exact-cap test's second block is the discriminating case for the stderr arm's `>` vs `>=`. Nothing blocking.

### Resolved since last review

- `src-tauri/src/agent.rs:3200-3211` — the stderr arm now has independent boundary coverage (`capture_capped_truncates_and_flags_over_cap_on_stderr`), and `capture_capped_allows_output_exactly_at_cap` (`:3213-3229`) drives the exact-cap case on both streams, which is what actually kills a `>` → `>=` mutation on either arm.
- `src-tauri/src/agent.rs:751-755` — `run_capture_parts`'s doc now states the per-stream cap and the `AppError::Command` failure contract, matching the kill at `:781-783` and the overflow gate at `:789-797`. `run_capture` (`:806`) still delegates via `?`, so its one-liner needed no edit, and `MAX_CAPTURE_BYTES`'s doc stayed on the const.
- `src-tauri/src/agent.rs:794` — the overflow message reads "on stdout or stderr", so it no longer implies a total budget (CodeRabbit's remaining open comment; addressed).
- `src/features/pulls/PrReviewPanel.tsx:412-422` — the `ComboboxEmpty` copy now matches `AgentPickers.tsx:114-116` and `AiProviderSection.tsx:231-233` verbatim, and the amended comment is accurate: every `MODEL_SUGGESTIONS` entry is non-empty (`src/lib/ai/providers.ts:129-165`, plus `OPENAI_COMPATIBLE_PRESETS[0].models`), so the empty state is reachable only under a typed filter.

Recorded decisions 8 and 10 still hold on inspection: `changelog.d/added-opencode-live-models.md` is still pending, so #246 is unreleased and no released build carried either defect (this also answers Copilot's fragment comment), and the guide's only claims about this surface — `src/features/help/content.ts:2081` and `:2087-2089` — remain true, since HTTP providers still fetch eagerly and the opencode picker still fills from the CLI catalog once reached.

### Readability

- **nit** — `src/features/sessions/AgentPickers.tsx:114-116`, `src/features/settings/AiProviderSection.tsx:231-233`, `src/features/pulls/PrReviewPanel.tsx:419-421`: the same two user-facing strings are now hand-copied at three call sites, each with its own near-identical explanatory comment. Export one helper beside the hooks that own the contract — e.g. in `src/lib/ai/models.ts`, `export function modelPickerEmptyText(isFetching: boolean)` returning the two strings — carry the "placeholder suggestions are always present, so this renders only under a typed filter" rationale once in that helper's doc comment, and drop the three per-site comments when you replace their bodies with `{modelPickerEmptyText(available.isFetching)}` (`availableModels.isFetching` in `AiProviderSection`).

Nothing further to raise.

Verdict: no blocking issues — remaining items are non-blocking; merge when ready

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._

**Comment by theBGuy** [comment]
Round-3 nit landed in the next push: `modelPickerEmptyText(isFetching)` now lives in `src/lib/ai/models.ts` beside the two hooks that own the contract, carrying the placeholder-suggestions rationale once in its doc; all three call sites (`AgentPickers`, `AiProviderSection`, `PrReviewPanel`) render `{modelPickerEmptyText(…)}` and dropped their per-site copies of the strings and comments. Verified: `tsc -b`, biome ci on all four touched files (the gate caught — and I fixed — one over-length line the first form produced), NUL-clean. No Rust change this round.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI review** · `opus` · automated

---

This round only extracts the duplicated `ComboboxEmpty` copy into `modelPickerEmptyText` in `src/lib/ai/models.ts` and swaps the three call sites onto it. The extraction is correct and complete: both hooks it documents do keep non-empty placeholder suggestions (`fallbackModels` resolves to `MODEL_SUGGESTIONS`/`OPENAI_COMPATIBLE_PRESETS[0].models`, every entry non-empty at `src/lib/ai/providers.ts:129-181`), all three sites now import the helper and dropped their local strings and comments, and no other consumer of those strings remains in the tree.

### Resolved since last review

- `src/features/sessions/AgentPickers.tsx:111`, `src/features/settings/AiProviderSection.tsx:228`, `src/features/pulls/PrReviewPanel.tsx:414` — the triplicated copy and its three near-identical comments are gone; the rationale now lives once on `modelPickerEmptyText` (`src/lib/ai/models.ts:357-366`), and the doc's claim ("both hooks keep placeholder suggestions present") checks out against `placeholderData` at `:303` and `:349`. Imports on all three files are the named form alongside the existing hook import, with no stray leftovers.

Reviewing the fix hunks as new code: no collateral — the helper is a pure function of `isFetching`, the call sites pass the right query's flag (`available.isFetching` in the two panels, `availableModels.isFetching` in Settings), and the Rust side is untouched this round (`capture_capped`'s `n > room` boundary, the overflow → `AppError::Command` gate at `agent.rs:789-797`, and the four `#[tokio::test]`s all still read correctly, with `stdin(Stdio::null())`/`stdout`/`stderr` piped at `:766-768` making the `expect("stdout piped")` unreachable).

Recorded decisions 8 and 10 still hold on inspection: `changelog.d/added-opencode-live-models.md` is still pending, so #246 is unreleased and neither defect was observable in a shipped build (this is also why Copilot's fragment comment doesn't apply — the `no-changelog` label itself isn't visible from the diff), and the site capability line (`site/src/data/capabilities.ts:601`) claims only the live catalog and free-typed ids, both still true after the deferral, so no README/site/help edit is owed.

Nothing further to raise.

Verdict: no blocking issues — remaining items are non-blocking; merge when ready

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._
